apps/dash-label-properties/app.py

```diff
@@ -0,0 +1,376 @@
+import pandas as pd
+
+import PIL
+from skimage import color, img_as_ubyte
+from plotly import colors
+from textwrap import dedent
+
+
+external_stylesheets = [dbc.themes.BOOTSTRAP]
+
+app = dash.Dash(__name__, external_stylesheets=external_stylesheets)
+server = app.server
+
+# Read the image that will be segmented
+filename = (
+    "https://upload.wikimedia.org/wikipedia/commons/a/ac/Monocyte_no_vacuoles.JPG"
+)
+img = io.imread(filename, as_gray=True)[:660:2, :800:2]
+labels = measure.label(img < filters.threshold_otsu(img))
+# Compute and store properties of the labeled image
+prop_names = [
+    "label",
+    "area",
+    "perimeter",
+    "eccentricity",
+    "euler_number",
+    "mean_intensity",
+]
+prop_table = measure.regionprops_table(
+    labels, intensity_image=img, properties=prop_names
+)
+table = pd.DataFrame(prop_table)
+# Format the Table columns
+columns = [
+    {"name": label_name, "id": label_name}
+    if precision is None
+    else {
+        "name": label_name,
+        "id": label_name,
+        "type": "numeric",
+        "format": Format(precision=precision),
+    }
+    for label_name, precision in zip(prop_names, (None, None, 4, 4, None, 3))
+]
+
+
+def image_with_contour(img, labels, data_table, mode="lines", shape=None):
+    """
+    Figure with contour plot of labels superimposed on background image.
+
+    Parameters
+    ----------
+
+    img : URL, dataURI or ndarray
+        Background image. If a numpy array, it is transformed into a PIL
+        Image object.
+    labels : 2D ndarray
+        Contours are the isolines of labels.
+    shape: tuple, optional
+        Shape of the arrays, to be provided if ``img`` is not a numpy array.
+    """
+    try:
+        sh_y, sh_x = shape if shape is not None else img.shape
+    except AttributeError:
+        print(
+            """the shape of the image must be provided with the
+                 ``shape`` parameter if ``img`` is not a numpy array"""
+        )
+    if type(img) == np.ndarray:
+        img = img_as_ubyte(color.gray2rgb(img))
+        img = PIL.Image.fromarray(img)
+    labels = labels.astype(np.float)
+    custom_viridis = colors.PLOTLY_SCALES["Viridis"]
+    custom_viridis.insert(0, [0, "#FFFFFF"])
+    custom_viridis[1][0] = 1.0e-4
+    # Contour plot of segmentation
+    print("mode is", mode)
+    opacity = 0.4 if mode is None else 1
+
+    # Compute overlay data to display in hovertemplate
+    overlay_columns = data_table.columns[1:]
+    indices = data_table.label.values
+    values = np.zeros((indices.max() + 1, len(overlay_columns)))
+    values[indices, :] = data_table[overlay_columns].values
+    overlay_data = np.dstack(
+        [
+            values[:, col_id][labels.astype(int)]
+            for col_id, col in enumerate(overlay_columns)
+        ]
+    )
+    # Display hover data with precision if data is float
+    hover_string = [
+        f"{col}: %{{customdata[{col_id}]:.3f}}"
+        if data_table[col].dtype in (np.float,)
```

Comment:
This is basically what `np.array_str` does (but it adds these `[...]` things around the numbers), so it sort of seems the way to do it. The arrays I think are just numpy arrays underneath. I think it's ok that `data_table[col].dtype in "float"` float has the same result, because in numpy you can make an array like `np.zeros(N,dtype='float')` or `np.zeros(N,dtype=np.float)` and they will have the same type. That it doesn't check that it is one of those two types is a another story involving Python's type checking system in general... but I think we don't have to worry about that (we provided the data, haha).

Reply:
Thanks for the pointers @nicholas-esterer! I dug a little further because I didn't understand why `np.dtype("float") == "float"` evaluates `True` essentially. 
The way I understand it now is that 1) [numpy datatypes have a hierarchy](https://numpy.org/doc/stable/reference/arrays.scalars.html) where `np.float64` is a sub-datatype of `float` and 2) numpy array's have [a `__eq__` method](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.__eq__.html) that evaluates this hierarchy internally. So you can also do things like `np.dtype("float") == "double"` and so on...
Guess I learned something new today!

Reply:
I think to be super explicit here, I'll just go with [`np.issubdtype`](https://numpy.org/doc/stable/reference/generated/numpy.issubdtype.html) here to avoid any confusion

Reply:
Maybe you can get away with just using printf statements, does this do what you want?
```
>>> '%0.3f' % (1.23,)
'1.230'
>>> '%0.3f' % (1,)
'1.000'
```
1 is an int but has 0s after it, 1.23 is made to have an extra 0